net48 and net6

veochen-octopus opened the pull request:
# Background

- `net452` is end of support and is preventing us from getting security patches both in the framework itself and 3rd party libraries.
- We'll need to support Ubuntu 22.04 pretty soon so `net6` is also required.

# Results

<!-- Describe the result of the change -->

Fixes https://app.shortcut.com/octopusdeploy/story/12030/target-net-framework-4-8-in-tentacle

## Before

```xml
<TargetFrameworks>net452;netcoreapp3.1</TargetFrameworks>
```

## After

```xml
<TargetFrameworks>net48;net6.0</TargetFrameworks>
```

# How to review this PR

<!--
Describe how you want people to review the pull request.
Perhaps you just want an "in principal" review to prove an idea.
Perhaps you want specific people to test the resulting changes.
-->

Quality :heavy_check_mark:
<!-- Describe focus areas (if any): Review tests/ Exploratory testing/ Smoke testing? -->

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

veochen-octopus commented on source/Octopus.Tentacle/Security/Certificates/CertificateEncoder.cs:
```diff
@@ -186,7 +186,7 @@ static bool HasPrivateKey(X509Certificate2 certificate2)
             try
             {
-                return certificate2.HasPrivateKey && certificate2.PrivateKey != null;
+                return certificate2.GetRSAPrivateKey() != null;
```
This is the only code change I had to make. The `EphemeralKeySet` flag is inside a `#if NET472_OR_GREATER || NETCOREAPP || NETSTANDARD` condition, so it was always ignored on `net452` and `PersistKeySet` would be used. From `net472` we have `EphemeralKeySet` but it was erroring due to private key not being found. As per the MS docs on `EphemeralKeySet` support:

> Keys loaded in this manner are almost always loaded via Windows CNG. Therefore, callers must access the private key by calling extension methods, such as [cert.GetRSAPrivateKey()](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.rsacertificateextensions.getrsaprivatekey). The [X509Certificate2.PrivateKey](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2.privatekey#system-security-cryptography-x509certificates-x509certificate2-privatekey) property does not function.

I'm assuming `GetECDsaPrivateKey` doesn't come into play here. There's only reference to RSA.

veochen-octopus replied:
Also TODO: Add a comment that this _may_ be reverted to `.PrivateKey` once .NET framework is dropped entirely. Will use this change to trigger the final build before merge.

tothegills replied:
Does this TODO need action?

veochen-octopus replied:
Yeah that was confusing. Looking at it again, no action needed.

veochen-octopus commented on source/Octopus.Tentacle/Security/Certificates/CertificateEncoder.cs:
```diff
@@ -103,7 +103,7 @@ static X509Certificate2 FromBase64String(string? thumbprint, string certificateS
         }
         
-#if NET472_OR_GREATER || NETCOREAPP || NETSTANDARD
+#if NETCOREAPP || NETSTANDARD
```
This used to be bypassed as we were targeting net452, but breaks net48. With `EphemeralKeySet` the cert loses the private key once added to the key store. As per the doc
> Since the keys are not persisted to disk, certificates loaded with this flag are not good candidates to add to an X509Store.

veochen-octopus replied:
This essentially allows us to keep the same behaviour as `net452`

veochen-octopus commented:
Cloning the tentacle project in teamcity for the net48 build is a bit wonky for some reason. The branch filter doesn't work all the time. Thinking once the PR is ready, we update the main tentacle project in teamcity first to target net48, then get a green build and merge.

veochen-octopus commented:
I'll post net48 green builds here til we update the main build chain. Here's the current one - https://build.octopushq.com/buildConfiguration/TeamFireAndMotion_OctopusTentacleNet48_SensibleDefaultsChainFullChain/5905429?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

veochen-octopus commented on source/Octopus.Tentacle.Tests.Integration/Octopus.Tentacle.Tests.Integration.csproj:
```diff
@@ -14,25 +14,25 @@
+        <PackageReference Include="FluentAssertions" Version="6.7.0" />
+        <PackageReference Include="Assent" Version="1.8.2" />
+        <PackageReference Include="NUnit" Version="3.13.3" />
+        <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
```
Taking the opportunity to update _some_ test packages. No non-test dependencies were changed.

tothegills commented:
What kind of customer comms are we doing? Does that need to happen before this is merged?

veochen-octopus commented:
> What kind of customer comms are we doing? Does that need to happen before this is merged?

@tothegills I think @octokhor is handling the comms side of things. Given the change of plan, I'll piggyback the net6 changes onto this PR as well so we end up with only 1 major version bump.

veochen-octopus commented on source/Octopus.Tentacle.Tests/Configuration/Crypto/LinuxMachineKeyEncryptorFixture.cs:
```diff
@@ -70,7 +70,7 @@ class InMemoryCryptoKeyNixSource : ICryptoKeyNixSource
             public InMemoryCryptoKeyNixSource()
             {
-                var d = new RijndaelManaged();
+                var d = Aes.Create();
```
`RijndaelManaged` obsolete. All references changed to `Aes.Create`

veochen-octopus commented on source/Octopus.Tentacle/Certificates/CertificateGenerator.cs:
```diff
@@ -97,7 +97,7 @@ X509Certificate2 Generate(string fullName, bool exportable)
             {
                 store.Save(ms, exportpw.ToCharArray(), random);
-                var platformSpecificX509KeyStorageFlags = PlatformDetection.IsRunningOnMac ? X509KeyStorageFlags.DefaultKeySet : X509KeyStorageFlags.EphemeralKeySet;
+                var platformSpecificX509KeyStorageFlags = PlatformDetection.IsRunningOnNix ? X509KeyStorageFlags.EphemeralKeySet : X509KeyStorageFlags.DefaultKeySet;
```
Don't think `EphemeralKeySet` works with Windows at all. See other change.

veochen-octopus commented on source/Octopus.Tentacle/Security/AesEncryption.cs:
```diff
@@ -20,69 +20,62 @@ public AesEncryption(string password)
-                BlockSize = 128,
-                Key = key
-            };
+            var provider = Aes.Create();
```
Another `obsolete`. Updated to current standard.

veochen-octopus commented:
Not sure about the best way to update the teamcity build. There are quite a few changes to be made and it's hard to keep track of them. Thinking probably the easiest is to replace the old tentacle project with the new one and make sure everything has the correct names and ids, get a green build and merge.

veochen-octopus commented on docker/windows/Dockerfile:
```diff
@@ -20,13 +21,10 @@ LABEL \
-
-WORKDIR /tmp
-RUN msiexec /i Octopus.Tentacle.%BUILD_NUMBER%-x64.msi /qn /l*v Tentacle-Installation.log
+COPY _artifacts/tentacle "${INSTALLATION_FOLDER}"
```
This is the only way I've found that works with `COPY` and Windows paths that contain spaces, i.e. with an `ARG`. Open to suggestions.

veochen-octopus commented:
Latest green build - https://build.octopushq.com/buildConfiguration/TeamFireAndMotion_OctopusTentacleNet48_SensibleDefaultsChainFullChain/5962281?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandPull+Request+Details=true

tothegills commented:
It'd be worth creating an Octopus Server branch, updating the Tentacle version and running the Octopus Server test suite.

tothegills commented on installer/Octopus.Tentacle.Installer/Product.wxs:
```diff
@@ -1,80 +1,59 @@
-
+    <PropertyRef Id="NETFRAMEWORK45" />
+    <PropertyRef Id="NETFRAMEWORK40FULLINSTALLROOTDIR" />
+    <Condition Message="This application requires Microsoft .NET Framework 4.5 Runtime in order to run. Please install the .NET Framework and then run this installer again."><![CDATA[Installed OR NETFRAMEWORK45]]></Condition>
```
I don't think this message about .NET framework 4.5 is still true

veochen-octopus replied:
🤔 I think this was auto generated. Let me see where it got this information from.

veochen-octopus replied:
Can I please get some help on this? Don't want to dump too much time into learning Wix (can do that later). But from what I can see from [the manual](https://wixtoolset.org/documentation/manual/v3/customactions/wixnetfxextension.html), there's no good replacement for `NETFRAMEWORK45`  for `net48`. This is wix v3. Not sure about v4.

Also found a [snippet in their source code](https://github.com/wixtoolset/wix3/blob/c02e48ec301a60eba88a3b519d47e88eeaa4c978/src/ext/NetFxExtension/wixlib/NetFx48.wxs) that has both similar and very different settings. A little confused...

tothegills commented on source/Octopus.Tentacle.Tests/Configuration/InMemoryKeyValueStoreFixture.cs:
```diff
@@ -91,8 +91,9 @@ public void ComplexTypeGetsHandledCorrectly()
             var settings = store.TryGet<TestConfig[]>("Test");
-            settings.value.Single().SettingA.Should().Be("some value", "strings should get parsed");
-            settings.value.Single().SomethingElse.Should().Be(12, "ints should get parsed");
+            settings.value.Should().NotBeNull();
```
The null check is an important assertion

tothegills commented on source/Octopus.Tentacle/Communications/OctopusServerChecker.cs:
```diff
@@ -38,34 +38,40 @@ public string CheckServerCommunicationsIsOpen(Uri serverAddress, IWebProxy? prox
             Retry("Checking that server communications are open", () =>
             {
-#pragma warning disable DE0003
+                // TODO see if can deal with SYSLIB0014
```
Is this TODO still to do? What are the plans for doing it?

veochen-octopus replied:
Removing this TODO. Fix would be to properly inject an `HttpClient` (or the factory) and go async from there. Treating it as out of scope for now, as it might not be a local change. But important to address later on. Ideally make tentacle more async overall.

tothegills commented on source/Octopus.Tentacle/Versioning/AssemblyExtensions.cs:
```diff
@@ -9,13 +9,13 @@ public static class AssemblyExtensions
         {
             var fileVersionInfo = FileVersionInfo.GetVersionInfo(assembly.FullLocalPath());
-            return fileVersionInfo.FileVersion;
+            return fileVersionInfo.FileVersion ?? "";
```
Is "" a good backup value?

veochen-octopus replied:
Good question. We could return "0.0.0". It would be more friendly to the semantic version parser. But it would also hide the fact that there is no version. Thinking about the responsibility of this method, _should_ it know what a good default is? Maybe we should simply return a `string?` and let the consumer decide what they want to use as the default. For example

```csharp
// SemanticVersionInfo.cs
SemanticVersion = SemanticVersion.Parse(assembly.GetInformationalVersion() ?? "0.0.0");

// OctopusTentacle.cs
public static readonly string InformationalVersion = AssemblyExtensions.GetInformationalVersion(Assembly) ?? "Unknown";
```

tothegills reviewed (commented):
Great work, way to move this forward.
I think there are some installer pre-reqs that need updating.
I think Fixes should be a public issue in this repository.
It would be great to have a green Octopus Server build with these changes integrated.

veochen-octopus commented:
Main changes

- Updated `AssemblyExtensions.GetInformationalVersion` to return a `string?` and force the consumer to provide a default, i.e. `assembly.GetInformationalVersion() ?? "0.0.0"`
- Updated `Product.wxs` to require net48. (Couldn't merge your PR Shane so I've copied your change)
- Enabled Ubuntu 22.04 test and added `libssl3` to the deb dependencies so that the test is passing
- Renamed new Teamcity build to `Octopus Tentacle 2022` and updated build ids.
  - Need to make sure the build number is valid before switching to the new build

tothegills commented on build/Build.Tests.cs:
```diff
@@ -85,8 +85,7 @@ void RunLinuxPackageTestsFor(TestConfigurationOnLinuxDistribution testConfigurat
                 new TestConfigurationOnLinuxDistribution(NetCore, "linux-x64", "linuxmintd/mint19.3-amd64", "deb"),
-                // new TestConfigurationOnLinuxDistribution(NetCore, "linux-x64", "ubuntu:latest", "deb"), // 22.04 doesn't support netcore, https://github.com/dotnet/core/issues/7038
-                // new TestConfigurationOnLinuxDistribution(NetCore, "linux-x64", "ubuntu:rolling", "deb"), // 22.04 doesn't support netcore, https://github.com/dotnet/core/issues/7038
+                new TestConfigurationOnLinuxDistribution(NetCore, "linux-x64", "ubuntu:jammy", "deb"), // 22.04
```
It's helpful to test against `latest` and `rolling` because they alert us to problems that are not caught by testing the explicit versions.

veochen-octopus replied:
I've added these back and turned them on. I assume when this starts to fail, we'll disable them again until we're ready for the next upgrade?

tothegills reviewed (commented):
Looks good to me. Intending to merge this PR on November 21st.

veochen-octopus commented:
Ah yes, the "something else" I was trying to remember in the discussion points from earlier is this PR https://github.com/OctopusDeploy/OctopusTentacle/pull/405. I did the same thing without knowing about it. Seeing there's an ongoing discussion in the other PR this could change, which we can do afterwards.
